proposals/p3848.md

```diff
@@ -0,0 +1,638 @@
+continuous syntax was the better path forward, even though the minimal lambda
+syntax would be a bit less terse.
+
+## Introducer
+
+**Proposal**: Introduce with the `fn` keyword to mirror function declarations.
+If a statement or declaration begins with `fn`, a name is required and it
+becomes a function declaration. Otherwise, if in an expression context, `fn`
+introduces a lambda.
+
+```
+let lambda1: auto = fn => T.Make();
+
+let lambda2: auto = fn []() -> T { return T.Make(); };
+
+fn FunctionDeclaration1 => T.Make();
+
+fn FunctionDeclaration2[]() -> T { return T.Make(); }
+```
+
+### Alternative Considered: Sigil
+
+Introduce with a sigil, such as `$` or `@`. Since introducer punctuation is such
+a scarce resource, and since there was no consensus on what sigil would best
+represent a lambda, and since there was a desire to create a more continuous
+syntax between lambdas and function declarations, this alternative was decided
+against.
+
+```
+let lambda1: auto = @ => T.Make();
+
+let lambda2: auto = @[]() -> T { return T.Make(); };
+```
+
+## Positional Parameters
+
+**Proposal**: Positional parameters, introduced in the body of a function by way
+of the dollar sign and a corresponding non-negative parameter position integer
+(ex: `$3`), are `auto` parameters to the function in which they are defined.
+They can be used in any lambda or function declaration that lacks an explicit
+parameter list (parentheses). They are variadic by design, meaning an unbounded
+number of arguments can be passed to any function that lacks an explicit
+parameter list. Only the parameters that are named in the body will be read
+from, meaning the highest named parameter denotes the minimum number of
+arguments required by the function. The function body is free to omit
+lower-numbered parameters (ex: `fn { Print($10); }`).
+
+```
+// A lambda that takes two positional parameters being used as a comparator
+Sort(my_list, fn => $0 < $1);
+```
+
+### Positional Parameter Restrictions
+
+**Proposal**: There are two restrictions applied to functions with positional
+parameters. The first restriction is that the definitions of function
+declarations must be attached to the declarations. The second restriction is
+that positional parameters can only be used in a context where there is exactly
+one enclosing function without an explicit parameter list. For example...
+
+```
+fn Foo1 {
+  fn Bar1 {}  // ❌ Invalid: Foo1 is already using positional parameters
+}
+
+fn Foo2 {
+  Print($0);
+  fn Bar2 {}  // ❌ Invalid: Foo2 is already using positional parameters
+}
+
+fn Foo3 {
+  fn Bar3 {
+    Print($0);  // ❌ Invalid: Foo3 is already using positional parameters
+  }
+}
+
+fn Foo4() {
+  fn Bar4 {
+    Print($0);  // ✅ Valid: Foo4 has explicit parameters
+  }
+}
+
+fn Foo5 {
+  fn Bar5() {}  // ✅ Valid: Bar5 has explicit parameters
+}
+
+fn Foo6() {
+  my_list.Sort(
+    fn => $0 < $1  // ✅ Valid: Foo6 has explicit parameters
```

Comment:
```suggestion
    fn => ($0 < $1)  // ✅ Valid: Foo6 has explicit parameters
```

Operator precedence thing: is this intended to evaluate as `(fn => $0) < $1` or `fn => ($0 < $1)`? I suspect we might not want to define precedence between `fn =>` and `<` and, as a consequence, parentheses would be required to disambiguate. It doesn't seem like the proposal is addressing precedence, which seems fine, but just suggesting the edit in the example for clarity.